File: python/day2/solution.py
import dataclasses
import enum


def parse_input(input_path: str) -> list[str]:
    with open(input_path, "r+") as file:
        return file.readlines()


@dataclasses.dataclass
class GamePower:
    score: int
    oponent_label: str
    player_label: str
    win: list[str]
    lose: list[str]


SCISSORS = GamePower(3, "C", "Z", ["B", "Y"], ["A", "X"])
PAPER = GamePower(2, "B", "Y", ["A", "X"], ["C", "Z"])
ROCK = GamePower(1, "A", "X", ["C", "Z"], ["B", "X"])


class Weapons(enum.Enum):
    A = ROCK
    B = PAPER
    C = SCISSORS
    X = ROCK
    Y = PAPER
    Z = SCISSORS


def solution1(data: list[str]) -> int:
    total_score = 0
    for line in data:
        oponent, me = line.strip().split(" ")
        me = Weapons[me].value
        if oponent in me.win:
            total_score += 6 + me.score
        elif oponent == me.oponent_label:
            total_score += 3 + me.score
        else:
            total_score += me.score

    return total_score


def solution2(data: list[str]) -> int:
    total_score = 0
    for line in data:
        oponent, outcome = line.strip().split(" ")
        oponent = Weapons[oponent].value
        if outcome == "Z":
            total_score += Weapons[oponent.lose[0]].value.score + 6
        elif outcome == "X":
            total_score += Weapons[oponent.win[0]].value.score
        else:
            total_score += oponent.score + 3

    return total_score


if __name__ == "__main__":
    print("Day2")
    solution1(parse_input("../../inputs/day2.txt"))
    solution2(parse_input("../../inputs/day2.txt"))

File: ruby/day2/solution.rb
class GamePower
    attr_accessor :score, :oponent_label, :player_label, :win, :lose

    def initialize(score, oponent_label, player_label, win, lose)
        @score = score
        @oponent_label = oponent_label
        @player_label = player_label
        @win = win
        @lose = lose
    end
end

class Weapons

    SCISSORS = GamePower.new(3, "C", "Z", Array["B", "Y"], ["A", "X"])
    PAPER = GamePower.new(2, "B", "Y", ["A", "X"], ["C", "Z"])
    ROCK = GamePower.new(1, "A", "X", ["C", "Z"], ["B", "X"])

    WEAPONS = Hash["A" => ROCK, "B" => PAPER, "C" => SCISSORS, "X" => ROCK, "Y" => PAPER, "Z" => SCISSORS]
end

def solution1(data)
    total_score = 0
    data.each do |line|
        oponent, me = line.strip().split(" ")
        me = Weapons::WEAPONS[me]
        if me.win.include?(oponent)
            total_score += 6 + me.score
        elsif oponent == me.oponent_label
            total_score += 3 + me.score
        else
            total_score += me.score
        end
    end

    return total_score
end

def solution2(data)
    total_score = 0
    data.each do |line|
        oponent, outcome = line.strip().split(" ")
        oponent = Weapons::WEAPONS[oponent]
        if outcome == "Z"
            total_score += Weapons::WEAPONS[oponent.lose[0]].score + 6
        elsif outcome == "X"
            total_score += Weapons::WEAPONS[oponent.win[0]].score
        else
            total_score += oponent.score + 3
        end
    end

    return total_score
end

def parse_input(path)
    file = File.open(File.join(File.dirname(__FILE__), path))
    return file.readlines
end

puts "Score exploited is: %d" % solution1(parse_input("../../inputs/day2.txt"))
puts "Planned score is: %d" % solution2(parse_input("../../inputs/day2.txt"))